### PR TITLE
KAA-1166: Add a page on configuration management

### PR DESCRIPTION
**Pull request by v-chyzhevskyi**
Page of 'Configuration management' fixed.

### CODE REVIEW

**Comment by ashvayka** [comment]
LGTM

File: doc/Programming-guide/Key-platform-features/Configuration-management/index.md [hunk 1]
```diff
@@ -5,51 +5,115 @@ permalink: /:path/
...
-- [Schema versioning](#schema-versioning)
-
-
-The Kaa Configuration subsystem is responsible for configuring endpoints by supplying them with the structured data of user-defined complexity that is managed via the Kaa server. The fact that Kaa operates with the uniformly structured data deserves a special emphasis. By knowing the data structure of the application, Kaa provides a number of very useful features, such as:
-
-* Incremental data updates for endpoints
-* Endpoint-specific data view that is based on the endpoint membership in endpoint groups
-* Automatic generation of the data object model in the endpoint SDK
+{% assign root_url = page.url | split: '/'%}
+{% capture root_url  %} /{{root_url[1]}}/{{root_url[2]}}/{% endcapture %}
```
Comment (balldir) [inline doc/Programming-guide/Key-platform-features/Configuration-management/index.md]: deprecated

File: doc/Programming-guide/Key-platform-features/Configuration-management/index.md [hunk 2]
```diff
@@ -437,312 +544,383 @@ The base schema is obtained by transforming the configuration schema, as follows
...
+#include <kaa_configuration_manager.h>
+ 
+static kaa_client_t *kaa_client = NULL;
 
-* The \__uuid field was added to org.kaaproject.sample.addressableRecordT, but not to org.kaaproject.sample.primitiveRecordT.
-* There is a reference to org.kaaproject.sample.addressableRecordT in the union definition for the delta field of the org.kaaproject.configuration.deltaT type.
-* org.kaaproject.configuration.uuidT is used in a union to indicate the removal of UUID-addressable records in /arrayOfRecords, but not in /arrayOfPrimitives.
+void print_configuration(
+        const kaa_root_configuration_t *configuration) {
+        printf("Current configuration: " configuration);
```
Comment (balldir) [inline doc/Programming-guide/Key-platform-features/Configuration-management/index.md:813]: > printf("Current configuration: " configuration);

Will fail at compile time

File: doc/Programming-guide/Key-platform-features/Configuration-management/index.md [hunk 3]
```diff
@@ -437,312 +544,383 @@ The base schema is obtained by transforming the configuration schema, as follows
...
+                                            const kaa_root_configuration_t *configuration) {
+    (void) context;
+    printf("Received configuration data\r\n");
+    print_configuration(configuration);
+    kaa_client_stop(kaa_client);
+    return KAA_ERR_NONE;
+}
 
-The delta calculation algorithm is executed by the Operations server to discover and represent via the [protocol schema]() the difference between the newly required configuration and the current one in regard to the specific endpoint. This difference is represented according to the corresponding [protocol schema]() and then supplied to the endpoint as the _delta configuration data_ (other accepted terms for which are delta, delta update, delta record). Note that both configurations should conform to the same configuration schema version.
+int main(/*int argc, char *argv[]*/) {
```
Comment (balldir) [inline doc/Programming-guide/Key-platform-features/Configuration-management/index.md:825]: > int main(/_int argc, char *argv[]_/) {

 Commented out code in the example is a bad thing.

File: doc/Programming-guide/Key-platform-features/Configuration-management/index.md [hunk 4]
```diff
@@ -437,312 +544,383 @@ The base schema is obtained by transforming the configuration schema, as follows
...
 
-To keep track of the current endpoint configuration, the Kaa server maintains a hash look-up table that contains the configuration SHA-1 hash, as well as the configuration data itself. Whenever the Operations server calculates a configuration with the unique hash, it adds that configuration to the table which is persisted in the database and shared across the Operations servers. The endpoint submits the hash of the last known configuration to Operations server for the table lookup.
+    //Initialize Kaa client.
+    kaa_error_t error_code = kaa_client_create(&kaa_client, NULL);
+    if (error_code) {
+        printf("Failed create Kaa client\r\n");
+        return 2;
```
Comment (balldir) [inline doc/Programming-guide/Key-platform-features/Configuration-management/index.md:831]: what does 2 mean ?

File: doc/Programming-guide/Key-platform-features/Configuration-management/index.md [hunk 5]
```diff
@@ -437,312 +544,383 @@ The base schema is obtained by transforming the configuration schema, as follows
...
-The described traversal method works for all [addressable fields]() in the configuration. However, items in the arrays are not addressable, therefore Kaa employs special handling logic for calculating deltas for the array fields, as follows:
+    error_code = kaa_configuration_manager_set_root_receiver(
+            kaa_client_get_context(kaa_client)->configuration_manager,
+            &receiver);
 
-* If the array items are UUID-addressable, the algorithm looks for the items with matching UUIDs in the two configuration sets and creates a separate delta for these items.
-* If the new configuration set misses some of the items, the algorithm proceeds as follows according to the item type:
+    if (error_code) {
+        printf("Failed to add configuration receiver\r\n");
+        return 3;
```
Comment (balldir) [inline doc/Programming-guide/Key-platform-features/Configuration-management/index.md:845]: what does 3 mean ?

**Comment by rasendubi** [comment]
C/C++/Obj-C examples are not valid. @chv348 you're not qualified enough to implement them properly, so please _don't_ do that or request proper review.

File: doc/Programming-guide/Key-platform-features/Configuration-management/index.md [hunk 6]
```diff
@@ -437,312 +544,383 @@ The base schema is obtained by transforming the configuration schema, as follows
...
+    //Initialize Kaa client.
+    kaa_error_t error_code = kaa_client_create(&kaa_client, NULL);
+    if (error_code) {
+        printf("Failed create Kaa client\r\n");
+        return 2;
+    }
 
-Having found the current endpoint configuration via the hash look-up, the Operations server starts comparing the current configuration set with the new one to detect any discrepancies between the two. Once all the differences are found, the Operations server looks up the nearest UUID-addressable record in the [addressing hierarchy]() and creates a delta record for it. After that, the algorithm proceeds to comparing the corresponding records in the two data sets and filling in the delta record. In case some field has the same value in both configuration sets, this field is assigned the unchanged type value ("unchangedT") in the delta record to indicate so. Otherwise, the value from the new configuration set is copied into the delta record.
+    kaa_configuration_root_receiver_t receiver = {
+            NULL,
```
Comment (forGGe) [inline doc/Programming-guide/Key-platform-features/Configuration-management/index.md:835]: Use designated initializer for structure elements.

File: doc/Programming-guide/Key-platform-features/Configuration-management/index.md [hunk 7]
```diff
@@ -437,312 +544,383 @@ The base schema is obtained by transforming the configuration schema, as follows
...
+int main(/*int argc, char *argv[]*/) {
 
-To keep track of the current endpoint configuration, the Kaa server maintains a hash look-up table that contains the configuration SHA-1 hash, as well as the configuration data itself. Whenever the Operations server calculates a configuration with the unique hash, it adds that configuration to the table which is persisted in the database and shared across the Operations servers. The endpoint submits the hash of the last known configuration to Operations server for the table lookup.
+    //Initialize Kaa client.
+    kaa_error_t error_code = kaa_client_create(&kaa_client, NULL);
+    if (error_code) {
+        printf("Failed create Kaa client\r\n");
```
Comment (forGGe) [inline doc/Programming-guide/Key-platform-features/Configuration-management/index.md:830]: Why `\r\n`?

**Pull request merged**